src/realm/object_converter.cpp

```diff
@@ -253,29 +244,335 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
+        }
+    }
+}
+
+void InterRealmValueConverter::handle_dictionary_in_mixed(Dictionary& src_dictionary,
+                                                          Dictionary& dst_dictionary) const
+{
+    std::vector<size_t> to_insert, to_delete;
+    size_t src_ndx = 0, dst_ndx = 0;
+    while (src_ndx < src_dictionary.size() && dst_ndx < dst_dictionary.size()) {
+        const auto [key_src, src_any] = src_dictionary.get_pair(src_ndx);
+        const auto [key_dst, dst_any] = dst_dictionary.get_pair(dst_ndx);
+
+        auto cmp = key_src.compare(key_dst);
+        if (cmp == 0) {
+            if (src_any != dst_any) {
+                to_insert.push_back(src_ndx);
+            }
+            else if (is_collection(src_any) &&
+                     !check_matching_dictionary(src_dictionary, dst_dictionary, key_src.get_string(),
+                                                to_collection_type(src_any))) {
+                to_insert.push_back(src_ndx);
+            }
+            src_ndx += 1;
+            dst_ndx += 1;
+        }
+        else if (cmp < 0) {
+            to_insert.push_back(src_ndx);
+            src_ndx += 1;
+        }
+        else {
+            to_delete.push_back(dst_ndx);
+            dst_ndx += 1;
+        }
+    }
+
+    // append src to dst
+    while (src_ndx < src_dictionary.size()) {
+        to_insert.push_back(src_ndx);
+        src_ndx += 1;
+    }
+
+    // delete everything that did not match passed src.size()
+    while (dst_ndx < dst_dictionary.size()) {
+        to_delete.push_back(dst_ndx);
+        dst_ndx += 1;
+    }
+
+    // delete all the non matching keys
+    while (!to_delete.empty()) {
+        dst_dictionary.erase(dst_dictionary.begin() + to_delete.back());
+        to_delete.pop_back();
+    }
+
+    // insert into dst
+    for (const auto pos : to_insert) {
+        const auto [key, any] = src_dictionary.get_pair(pos);
+        if (is_collection(any)) {
+            auto coll_type = to_collection_type(any);
+            dst_dictionary.insert_collection(key.get_string(), coll_type);
+            copy_dictionary_in_mixed(src_dictionary, dst_dictionary, key.get_string(), coll_type);
+        }
+        else {
+            dst_dictionary.insert(key, any);
+        }
+    }
+}
+
+bool InterRealmValueConverter::check_matching_list(const Lst<Mixed>& src_list, Lst<Mixed>& dst_list, size_t ndx,
+                                                   CollectionType type) const
+{
+
+    if (type == CollectionType::List) {
+        auto nested_src_list = src_list.get_list(ndx);
+        auto nested_dst_list = dst_list.get_list(ndx);
+        auto size_src = nested_src_list->size();
+        auto size_dst = nested_dst_list->size();
+        if (size_src != size_dst)
+            return false;
+        for (size_t i = 0; i < size_src; ++i) {
+            auto src_mixed = nested_src_list->get_any(i);
+            auto dst_mixed = nested_dst_list->get_any(i);
+            if (src_mixed != dst_mixed)
```

Comment:
Just to check that type of the mixed matches, the idea is that if the type matches, I defer the work of expanding the particular mixed (if it is representing a collection) later recursively. If the type does not match, there is no point in continuing forward, since the Mixed in destination will have to be replaced with whatever source has.

Reply:
Got it. This is new behaviour of Mixed equality that I wasn't familiar with.

Reply:
Yes, `Mixed` now carry the information about the type they represent. So we can do this type of comparisons for collections, obviously we can only compare the type (`List vs Set etc`)